lib/Sema/TypeCheckCompilerEvaluable.cpp

```diff
@@ -26,26 +26,205 @@ using namespace swift;
 
 namespace {
 
-/// Checks that the body of a function is compiler evaluable.
-/// Currently a skeleton implementation that only rejects while loops.
+/// Checks that a type is compiler representable.
+/// Currently a skeleton implementation that only rejects types named Float,
+/// Double and String.
 /// TODO(marcrasi): Fill in a real implementation.
+static bool checkCompilerRepresentable(const Type &type) {
+  return type.getString() != "Double" && type.getString() != "Float" &&
+         type.getString() != "String";
+}
+
+/// Checks that the body of a function is compiler evaluable.
 class CheckCompilerEvaluableBody : public ASTWalker {
   TypeChecker &TC;
-  bool compilerEvaluable = true;
 
- public:
-  CheckCompilerEvaluableBody(TypeChecker &TC) : TC(TC) {}
+  // The function whose body we are checking.
+  const AbstractFunctionDecl *CheckingFunc;
+
+  // Whether the body has passed the check.
+  bool CompilerEvaluable = true;
+
+public:
+  CheckCompilerEvaluableBody(TypeChecker &TC,
+                             const AbstractFunctionDecl *CheckingFunc)
+      : TC(TC), CheckingFunc(CheckingFunc) {}
+
+  std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
+    // If this is the ignored part of a DotSyntaxBaseIgnored, then we can accept
+    // it without walking it.
+    if (auto *parentDotSyntaxBaseIgnored =
+            dyn_cast_or_null<DotSyntaxBaseIgnoredExpr>(Parent.getAsExpr()))
```

Comment:
dyn_cast_or_null instead of dyn_cast is used heer because the input can be NULL -- what would be an example?

Reply:
when the parent is not an expression, it's null. for example, the top expression node for the condition of an `if` statement has the `if` statement as the parent.